In [1]:
from math import factorial

In [2]:
def combinations(n, k):
    return factorial(n) / (factorial(k) * factorial (n - k))

In [3]:
def arrangements(n, k):
    return factorial(n) / factorial (n - k)

### Задание №1
Из колоды в 52 карты извлекаются случайным образом 4 карты.  
a) Найти вероятность того, что все карты – крести.  
б) Найти вероятность, что среди 4-х карт окажется ХОТЯ БЫ один туз.  

### №1.а  
В колоде 52/4 = 13 крестовых карт,  
определим сначала колличество сочетаний по 4 карты из них,  
при n = 13 и k = 4 получаем:  
C = 13!/(4!(13-4)!): 

In [4]:
combinations(13, 4)

715.0

Определим теперь колличество всех возможных сочетаний по 4 карты из колоды в 52 карты:  
n = 52,  
k = 4,  
C = 52!/4!(13-4)!):

In [5]:
combinations(52, 4)

270725.0

Вероятность того, что в извлеченных случайно из колоды 4 картах, все окажутся крести:  
#### P = m/n, где   
m = 715 - колличество благоприятных исходов,  
n = 270725 - колличество всех исходов,  
#### m/n = 715/270725 = 0.0026, или 0.26%

### №1.б 

Определим сначала колличество сочетаний 4х карт, в которых будет 1, 2, 3 или 4 туза

In [6]:
n = 4
t = [] #список для колличеств сочетаний из тузов по 1, 2, 3 и 4 карты
for k in range(1,5):
    c = combinations(n, k)
    print('При k={} и n={} c={}'.format(k, n, c))
    t.append(c)

При k=1 и n=4 c=4.0
При k=2 и n=4 c=6.0
При k=3 и n=4 c=4.0
При k=4 и n=4 c=1.0


Вычислим колличество возможных сочетаний из 1, 2 и 3х карт в колоде за вычетом тузов 

In [7]:
n = 48
non_t = [] #список для колличеств сочетаний не тузов из 1, 2 и 3х карт
for k in range(1, 4):
    c = combinations(n, k)
    print('При k={} и n={} c={}'.format(k, n, c))
    non_t.append(c)

При k=1 и n=48 c=48.0
При k=2 и n=48 c=1128.0
При k=3 и n=48 c=17296.0


Найдем теперь колличество сочетаний:  
* один туз + 3 не туза
* два туза + 2 не туза
* три туза + 1 туз
* 4 туза

In [8]:
non_t.reverse()
non_t.append(1) #случай, когда все 4 карты - тузы
for c_t, c_non_t in zip(t, non_t):
    print('{} x {} = {}'.format(c_t, c_non_t, c_t*c_non_t))


4.0 x 17296.0 = 69184.0
6.0 x 1128.0 = 6768.0
4.0 x 48.0 = 192.0
1.0 x 1 = 1.0


Вычислим теперь число всех благоприятных исходов:

In [9]:
69184 + 6768 + 192 + 1

76145

Вычислим вероятность того, что среди 4х произвольных карт хотя бы одна окажется тузом,  
число всех возможных исходов возьмем из решения задания 1а

Вероятность того, что из 4х произвольных карт одна окажется тузом:  

In [10]:
m = 76145  
n = 270725
print('P = {:.4f} = {:.2f}%'.format(m/n, m/n*100))

P = 0.2813 = 28.13%


### Задание №2
На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9. Код содержит три цифры, которые нужно нажать одновременно. Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

Для каждого из 3х разрядов замка  
колличество возможных исходов n = 10  
колличество благоприятных исходов m = 1  
вероятность верно ввести одну цифру 1/10  
Угадывание каждой из 3х цифр по отдельности - независимое от других цифр событие  
Вероятность наступления независимых событий определяется как произведение их вероятностей, соответственно  
P(ABC) = P(A) x P(B) x P(C) = 1 / 10**3 = 0.001
### P = 0.001 = 0.1%

### Задание №3
В ящике имеется 15 деталей, из которых 9 окрашены. Рабочий случайным образом извлекает 3 детали. Какова вероятность того, что все извлеченные детали окрашены?

Вероятность того, что первая деталь окажется окрашенной:  
m = 9, n = 15, 
P(A) = m/n = 9/15,  
вероятность того, что вторая деталь окажется окрашенной при условии, что окрашена первая  
m = 8, n = 14  
P(B|A) = 8/14  
вероятность того, что из 2х извлеченных деталей обе окрашены, как вероятность двух зависимых событий    
P(A * B) = P(A) * P (B|A) = (9/15) * (8/14)  
аналогично для третей детали:  
m = 7, n = 13  
P(C| A * B) = 7/13  
вероятность того, что все 3 детали окрашены:  
P(A * B * C) = (9/15) * (8/14) * (7/13) = 504/2730 =   

In [11]:
print('P = {:.4f} = {:.2f}%'.format(504/2730, 504/2730*100))

P = 0.1846 = 18.46%


Вариант решения спомощью сочетаний:
Все возможные сочетания из 3х деталей для 15 деталей  
С = 15!/(3!(15-3)!)  
Все возможные сочетания из 3х деталей для 9 окрашенных деталей  
С = 9!/(3!(9-3)!)

In [12]:
С = combinations(9, 3) / combinations(15, 3)
print('P = {:.4f} = {:.2f}%'.format(С, С*100))

P = 0.1846 = 18.46%


#### Задание №4
В лотерее 100 билетов. Из них 2 выигрышных. Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

Вариант решения спомощью сочетаний:  
Все возможные сочетания из 2х билетов для 100 билетов  
С = 100!/(2!(100-2)!) = (98! * 99 * 100) / (1 * 2) * 98! = (99 * 100) / 2 = 4950  
Все возможные сочетания из 2х билетов для 2 выигрышных билетов  
С = 2!/(2!(2-2)!) = 2 /(2 * 1) = 1  

In [15]:
P = 1/4950

In [16]:
print('P = {:.4f} = {:.2f}%'.format(P, P*100))  


P = 0.0002 = 0.02%
